AttributeError: 'types.UnionType' object has no attribute '__name__'

Hi there! It seems like your library does not support Union types.

## Working
```python
@dataclass(frozen=True)
class Bar:
    foo: Optional[str] = None
```

## Not working
```python
@dataclass(frozen=True)
class Bar:
    foo: str | None = None
```

## Error message:
```
/home/wsontopski/Documents/Gitlab/maddox/backend/src/shared/nats/serialization.py:102: in _get_marshmallow_schema_of_dataclass
    return marshmallow_dataclass.class_schema(clazz=class_obj)
/home/wsontopski/.cache/pypoetry/virtualenvs/maddox-E1xGxQ_g-py3.10/lib/python3.10/site-packages/marshmallow_dataclass/__init__.py:356: in class_schema
    return _internal_class_schema(clazz, base_schema, clazz_frame)
/home/wsontopski/.cache/pypoetry/virtualenvs/maddox-E1xGxQ_g-py3.10/lib/python3.10/site-packages/marshmallow_dataclass/__init__.py:402: in _internal_class_schema
    attributes.update(
/home/wsontopski/.cache/pypoetry/virtualenvs/maddox-E1xGxQ_g-py3.10/lib/python3.10/site-packages/marshmallow_dataclass/__init__.py:405: in <genexpr>
    field_for_schema(
/home/wsontopski/.cache/pypoetry/virtualenvs/maddox-E1xGxQ_g-py3.10/lib/python3.10/site-packages/marshmallow_dataclass/__init__.py:729: in field_for_schema
    or _internal_class_schema(typ, base_schema, typ_frame)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

clazz = str | None, base_schema = None
clazz_frame = <frame at 0x7fc340fc2c20, file '/home/wsontopski/Documents/Gitlab/maddox/backend/src/shared/nats/serialization.py', line 102, code _get_marshmallow_schema_of_dataclass>

    @lru_cache(maxsize=MAX_CLASS_SCHEMA_CACHE_SIZE)
    def _internal_class_schema(
        clazz: type,
        base_schema: Optional[Type[marshmallow.Schema]] = None,
        clazz_frame: types.FrameType = None,
    ) -> Type[marshmallow.Schema]:
>       _RECURSION_GUARD.seen_classes[clazz] = clazz.__name__
E       AttributeError: 'types.UnionType' object has no attribute '__name__'

/home/wsontopski/.cache/pypoetry/virtualenvs/maddox-E1xGxQ_g-py3.10/lib/python3.10/site-packages/marshmallow_dataclass/__init__.py:367: AttributeError

```

## Context
Python 3.10.3
Marshmallow-Dataclass 8.5.8

--- a/marshmallow_dataclass/__init__.py
+++ b/marshmallow_dataclass/__init__.py
@@ -495,8 +495,8 @@ def _field_for_generic_type(
     If the type is a generic interface, resolve the arguments and construct the appropriate Field.
     """
     origin = typing_inspect.get_origin(typ)
+    arguments = typing_inspect.get_args(typ, True)
     if origin:
-        arguments = typing_inspect.get_args(typ, True)
         # Override base_schema.TYPE_MAPPING to change the class used for generic types below
         type_mapping = base_schema.TYPE_MAPPING if base_schema else {}
 
@@ -557,38 +557,38 @@ def _field_for_generic_type(
                 ),
                 **metadata,
             )
-        elif typing_inspect.is_union_type(typ):
-            if typing_inspect.is_optional_type(typ):
-                metadata["allow_none"] = metadata.get("allow_none", True)
-                metadata["dump_default"] = metadata.get("dump_default", None)
-                if not metadata.get("required"):
-                    metadata["load_default"] = metadata.get("load_default", None)
-                metadata.setdefault("required", False)
-            subtypes = [t for t in arguments if t is not NoneType]  # type: ignore
-            if len(subtypes) == 1:
-                return field_for_schema(
-                    subtypes[0],
-                    metadata=metadata,
-                    base_schema=base_schema,
-                    typ_frame=typ_frame,
-                )
-            from . import union_field
-
-            return union_field.Union(
-                [
-                    (
-                        subtyp,
-                        field_for_schema(
-                            subtyp,
-                            metadata={"required": True},
-                            base_schema=base_schema,
-                            typ_frame=typ_frame,
-                        ),
-                    )
-                    for subtyp in subtypes
-                ],
-                **metadata,
+    if typing_inspect.is_union_type(typ):
+        if typing_inspect.is_optional_type(typ):
+            metadata["allow_none"] = metadata.get("allow_none", True)
+            metadata["dump_default"] = metadata.get("dump_default", None)
+            if not metadata.get("required"):
+                metadata["load_default"] = metadata.get("load_default", None)
+            metadata.setdefault("required", False)
+        subtypes = [t for t in arguments if t is not NoneType]  # type: ignore
+        if len(subtypes) == 1:
+            return field_for_schema(
+                subtypes[0],
+                metadata=metadata,
+                base_schema=base_schema,
+                typ_frame=typ_frame,
             )
+        from . import union_field
+
+        return union_field.Union(
+            [
+                (
+                    subtyp,
+                    field_for_schema(
+                        subtyp,
+                        metadata={"required": True},
+                        base_schema=base_schema,
+                        typ_frame=typ_frame,
+                    ),
+                )
+                for subtyp in subtypes
+            ],
+            **metadata,
+        )
     return None
 
 

--- a/tests/test_union.py
+++ b/tests/test_union.py
@@ -1,4 +1,5 @@
 from dataclasses import field
+import sys
 import unittest
 from typing import List, Optional, Union, Dict
 
@@ -180,3 +181,18 @@ class TestClassSchema(unittest.TestCase):
             )
         with self.assertRaises(marshmallow.exceptions.ValidationError):
             schema.load({"value": None})
+
+    @unittest.skipIf(sys.version_info < (3, 10), "No PEP604 support in py<310")
+    def test_pep604_union(self):
+        @dataclass
+        class PEP604IntOrStr:
+            value: int | str
+
+        schema = PEP604IntOrStr.Schema()
+        data_in = {"value": "hello"}
+        loaded = schema.load(data_in)
+        self.assertEqual(loaded, PEP604IntOrStr(value="hello"))
+        self.assertEqual(schema.dump(loaded), data_in)
+
+        data_in = {"value": 42}
+        self.assertEqual(schema.dump(schema.load(data_in)), data_in)

